tools/schemapi/schemapi.py

```diff
@@ -334,6 +384,109 @@ def _get_leaves_of_error_tree(
+"""Key function for ``(json_path, ValidationError)``."""
+_fn_validator = cast("Callable[[ValidationError], str]", operator.attrgetter("validator"))  # fmt: off
+"""Key function for ``ValidationError.validator``."""
+
+
+def _lazy_subset_to_most_specific_json_paths(
+    json_path_errors: Iterator[tuple[str, ValidationError]], /
+) -> Iterator[Iterable[ValidationError]]:
+    """
+    Currently using a `list`, but typing it more restrictive to see if it can be avoided.
+
+    - Needs to be sorted to work with groupby
+    - Reversing allows prioritising more specific groups, since they are seen first
+    - Then re-reversed, to keep seen order
+
+    """
+    rev_sort = sorted(json_path_errors, key=_fn_path, reverse=True)
+    keeping: dict[str, Iterable[ValidationError]] = {}
+    for unique_path, grouped_errors in groupby(rev_sort, key=_fn_path):
+        if any(seen.startswith(unique_path) for seen in keeping):
+            continue
+        else:
+            keeping[unique_path] = [err for _, err in grouped_errors]
+    yield from reversed(keeping.values())
+
+
+def _lazy_deduplicate_errors(
+    grouped_errors: Iterator[Iterable[ValidationError]], /
+) -> Iterator[ValidationError]:
+    for element_errors in grouped_errors:
+        for validator, errors in groupby(
+            sorted(element_errors, key=_fn_validator), key=_fn_validator
+        ):
+            if validator == "additionalProperties":
+                errors = _lazy_additional_properties(errors)
+            elif validator == "enum":
+                errors = _lazy_enum(errors)
+            yield from _lazy_unique_message(errors)
+
+
+def _lazy_unique_message(
+    iterable: Iterable[ValidationError], /
+) -> Iterator[ValidationError]:
+    seen = set()
+    for el in iterable:
+        if el.message not in seen:
+            seen.add(el.message)
+            yield el
+
+
+def _lazy_additional_properties(
+    iterable: Iterable[ValidationError], /
+) -> Iterator[ValidationError]:
+    it = iter(iterable)
+    first = next(it)
+    if (
+        parent := cast("ValidationError", first.parent)
+    ) and parent.validator == "anyOf":
+        yield min(_rechain(first, it), key=lambda x: len(x.message))
+    else:
+        yield first
+
+
+def _lazy_enum(iterable: Iterable[ValidationError], /) -> Iterator[ValidationError]:
+    """
+    Temporary reusing the eager version to isolate issues.
+
+    The 3 errors rule applies per group.
+    """
+    # FIXME: Too simple
+    # Need to do an eager pass, as this skips intersections of non-overlapping enums
+    # yield reduce(_enum_inner, iterable)
+    yield from _deduplicate_enum_errors(list(iterable))
+
+
+def _enum_inner(prev: ValidationError, current: ValidationError, /) -> ValidationError:
+    """**Disabled**."""
+    longest = set(cast("list[str]", prev.validator_value))
+    contender = set(cast("list[str]", current.validator_value))
+    return current if contender.issuperset(longest) else prev
```

Comment:
# Note
Originally added to use like below.
However this was naive, due to this skipping disjoint enums.

```py
from functools import reduce

def _lazy_enum(iterable: Iterable[ValidationError], /) -> Iterator[ValidationError]:
    yield reduce(_enum_inner, iterable)
```

# Possible alternative
Rather than preserving each enum, we could just extract the intersection of all of them.

It would produce a different error message than the original, so for the below:

https://github.com/vega/altair/blob/591bf40fc0158ddd8d2fad2774d5230eb73ab734/tests/utils/test_schemapi.py#L779-L788

We could adapt the formatting from [`SchemaValidationError._format_params_as_table`](https://github.com/vega/altair/blob/591bf40fc0158ddd8d2fad2774d5230eb73ab734/tools/schemapi/schemapi.py#L660) to print something like this:

<details><summary>Large Screenshot</summary>
<p>


![image](https://github.com/user-attachments/assets/d3c39a73-20f0-4dc2-aee0-0a51d99f8b7b)

</p>
</details> 

Which would just read as:

> 'invalid_value' is an invalid value for `timeUnit`. Valid values are: 
> - One of [...]
> - Of type 'object'

This would also be easier to control the formatting than relying on the groups provided by the enums alone